feat(discover): Prevent calling Snuba with an empty list of projects

armenzg opened the pull request:
[project_id](https://github.com/getsentry/snuba/blob/9f421e25256613fb003147a871f7adcbc248bd08/snuba/datasets/configuration/events/entities/events.yaml#L451-L454) is a required column for the errors entity, however, passing the empty list of IDs does not complain.

This change will allow the Sentry developer to see an error while doing development rather than debugging why a Snuba query is resulting in no data.

armenzg committed:
feat(discover): Prevent calling Snuba with an empty list of projects

gggritso committed:
fix(perf): Add optional chaining to list widget checks (#69497)

These values are sometimes `undefined` at runtime. Fixes JAVASCRIPT-2SRE

markstory committed:
feat(actor) Start dual writing for Rule (#69427)

Write to both the `owner` and `owner_user_id` and `owner_team_id`
columns. Part of the work to remove `Actor`.

Refs HC-1169

shruthilayaj committed:
feat(process-spans): Batch write spans to redis (#69445)

1. multiprocesser step deserializes spans to extract segment id
2. batch step collect spans and batch writes to redis instead of once
per message
3. offsets are committed per batch processed

DominikB2014 committed:
feat(cache): add hit/miss chart and throughput graph (#69455)

![image](https://github.com/getsentry/sentry/assets/44422760/2d75451c-632d-4ade-81ad-6c44f061bebd)

All landing page work
1. Add hit/miss graph
2. Add throughput graph
3. Add date filters, env selector, proj selector

armenzg commented:
FYI I'm looking to fix the tests.

colin-sentry committed:
feat(ai): Add a "total tokens used" readout to AI analytics (#69502)

armenzg committed:
Fix the tests

armenzg committed:
ref(snuba): Drop bulk_snql_query (#69402)

colin-sentry committed:
feat(ai): Add a ai_total_tokens_used span metric function (#69503)

armenzg committed:
Add tests

armenzg committed:
Fix typing

armenzg commented on tests/sentry/api/endpoints/test_organization_dashboard_widget_details.py:
```diff
@@ -638,6 +638,8 @@ def test_save_with_invalid_orderby_not_from_columns_or_aggregates(self):
 
     def test_save_with_total_count(self):
+        # We cannot query the Discover entity without a project being defined for the org
+        self.create_project()
```
If there's at least one project in the organization, we will use the list of projects matching `project_id`:
https://github.com/getsentry/sentry/blob/85558cb8f2b11fad6d74f048049ec33824624644/src/sentry/search/events/builder/discover.py#L134

JoshFerge committed:
feat(feedback): dont send notifications for spam messages (#69478)

gggritso committed:
feat(perf): Add HTTP module unknown domain banner (#69505)

Unknown domains are an annoying and fixable error. I don't want people
to just give up when they see this! It might be a matter of updating
their SDK, or something else simple.

malwilley committed:
fix(useoverlay): Enable closest overlay behavior for any element type (#69474)

This logic ensures that, when an overlay is open, you don't need to
click twice to open a different overlay.

ceorourke committed:
ref(rules): Don't catch snuba query rate limit exceeded errors for now (#69467)

We're hitting the abuse limit because this error is happening so
frequently so for now don't raise it, but emit a metric we can track
while the work to address the issue is underway.

narsaynorath committed:
feat(mobile-starfish): Add platform compatibility checker (#69432)

Adds a component that triggers a `spansIndexed` request to fetch data to
validate that the module being viewed has support for the data of the
projects selected.

This is used by the mobile modules because we've
noticed traffic from non-supported platforms and we need to more clearly
communicate to the users that they are viewing the module without the
correct projects selected.

scttcper committed:
fix(performance): Save query state earlier (#69477)

malwilley committed:
feat(query-builder): Add ability to set filter values (#69476)

Closes https://github.com/getsentry/sentry/issues/69263

- Added `focus` state to the reducer which tracks the currently focused
portion of the query
- Added actions for clicking into the token, updating the value, and
exiting the token (which clears focus)
- Added components to render the editing value state as a combobox

armenzg commented on src/sentry/search/events/builder/discover.py:
```diff
@@ -583,13 +583,17 @@ def resolve_params(self) -> list[WhereType]:
-                self.column("project_id"),
-                Op.IN,
-                self.params.project_ids,
+        # The clause will prevent calling Snuba with an empty list of projects, thus, returning
+        # no data. It will not instead complain with:
+        # sentry.utils.snuba.UnqualifiedQueryError: validation failed for entity events: missing required conditions for project_id
+        if self.params.project_ids:
```
@Zylphrex it will raise an error.
Not all datasets have `project_id` as a required column.

wedamija replied:
So we're relying on snuba to catch the error for us here?

Are there actually datasets we use in discover that don't require `project_id`?

armenzg replied:
That's the idea.

Most of them require it but some don't:
* [outcomes](https://github.com/getsentry/snuba/blob/cd1c90f72987611e422a42b9a185960e852bd60e/snuba/datasets/configuration/outcomes/entities/outcomes.yaml#L41)
* [outcomes raw](https://github.com/getsentry/snuba/blob/cd1c90f72987611e422a42b9a185960e852bd60e/snuba/datasets/configuration/spans/entities/spans.yaml#L161)

armenzg replied:
I tried using the query builder with outcomes and it is not supported.
I will look into raising an error instead.

scttcper committed:
feat(issues): Cleanup duplicate event states (#69481)

michellewzhang committed:
chore(replay/issues): add issue_replay_inline_onboarding to prompts config (#69513)

- Relates to https://github.com/getsentry/sentry/issues/69207
- Will be used to update the dismiss mechanism of the issue details
replay inline onboarding banner
- Added `project_id` as a required field since we want to store the
dismissal across project

<img width="863" alt="SCR-20240423-jmxe"
src="https://github.com/getsentry/sentry/assets/56095982/a9e1efc3-3325-4669-9a9b-d44dda66fcda">

c298lee committed:
feat(replays): Component Names in Rage & Dead Breadcrumbs (#69508)

Use component name in rage and dead breadcrumbs if available.

Relates to https://github.com/getsentry/sentry/issues/64673

cathteng committed:
chore(github): fix name of extensibility agent option and add an org option (#69514)

lobsterkatie committed:
ref(models): Pull title from `data` for default and error events/groups (#69397)

For both error and default (message) events, we calculate the event and group titles fairly early on in the ingest process, and store that information (among other places) in `event.data` and `group.data`. Nonetheless, whenever we ask for `event.title` or `group.title`, we recalculate the title via the event's `EventType` and its metadata, rather than just pulling it from `data`. 

This changes that, partially to save a few cycles and because it's a necessary prerequisite to fixing the bad title data left over from the bug last week wherein we were mistakenly classifying error events as default events, and preventing the same problem from happening again. (This is necessary for the prevention half of that fix because were the bug to recur, without this change we'd use the mistaken eventtype to get the title rather than the corrected title itself.) I chose only to apply the change to error and default events because figuring out if it was safe to apply more widely was out of scope for the data fix this PR enables. I left a TODO so we remember to investigate over event types.

Note: This change breaks a test I wrote way back when, which tests a warning we log when we don't have a good title to send seer. The logic being tested was written at the same time, back before I knew where in the ingest process we calculate titles, and so  I wrongly assumed recalculating the title before calling seer might help. Anyway, the point is, the change in this PR exposed the flaw in my logic, so for now, to keep things simple, I just xfailed the test. I'll put the actual fix into a follow-up PR. [UPDATE: Done in https://github.com/getsentry/sentry/pull/69450.]

azaslavsky committed:
fix(backup): Handle Rule changes during actor refactor (#69517)

Follow on for https://github.com/getsentry/sentry/pull/69227

wedamija committed:
chore(issue-platform): Stop relying on `group_states` in `post_process` (#63130)

`group_states` no longer needs to be used, because we'll only ever have
a single group at a time in `post_process`. I've verified this using
metrics added in https://github.com/getsentry/sentry/pull/62722.

Once we remove usage here, we can also remove code that sends
`group_states` through, and ideally also remove the multi group stuff
from `Event`.

iamrajjoshi committed:
chore(api-idorslug): Moved Path Param Tests (#69464)

Moved tests to `getsentry` - Merge
https://github.com/getsentry/getsentry/pull/13737 first.

leeandher committed:
ref(event-tags): Refactor work for Issue Details changes (#69426)

No significant changes here, but a lot of the util/refactors/new props
are necessary for enabling rendering custom highlights.

List of changes:
- Pulls out the list of context items from the `EventContexts` component
- Pulls the context content out of the `ContextCard` component (with a
config to edit how its rendered)
- Adds a config to `EventTagsTreeRow` component to edit how its rendered
- Remove a div from the `EventTagsTree`
- Creates the `ContextCard` components from the `EventContexts`
component into the `ContextDataSection`
- Implement a hook to get a detailed project from a slug (since
`useProjects` does not render with the `DetailedProjectSerializer`)
- Correct the flag name on some frontend tests
- Rename param for `useIssueDetailsColumnCount` hook
- Adds `highlightContext` and `highlightTags` to the Project type

evanpurkhiser committed:
feat(crons): Test the parallel threaded version of monitor_consumer (#69507)

Basically a follow up of https://github.com/getsentry/sentry/pull/64825,
but a little bit better.

---------

Co-authored-by: getsantry[bot] <66042841+getsantry[bot]@users.noreply.github.com>

gggritso committed:
fix(perf): Colour --> Color (#69520)

Whoopsie, eh

ceorourke committed:
ref(rules): Refactor `get_rate` to be used for bulk queries (#69446)

The `apply_delayed` rule processor needs to call `get_rate` with bulk
Snuba queries (see [this code
block](https://github.com/getsentry/sentry/blob/088b9c73b604743864f76bd39a81a3238699d306/src/sentry/rules/processing/delayed_processing.py#L136-L171)),
so this PR refactors `get_rate` to handle that.

0Calories committed:
feat(perf): Add feature flag for new spans UI (#69506)

Adds a feature flag for switching to the new UI for the span summary /
spans tab in txn summary. These pages have been updated with new designs
and leverage the span metrics dataset.

gggritso committed:
fix(perf): Improve `SlideOverPanel` layout (#69523)

- remove border
- add a nice box shadow
- clarify how the positioning is done
- prevent unnecessary scrollbars

iamrajjoshi committed:
chore(api): Updated Accept Organization Invite Endpoint (#69451)

The endpoint inherits straight from the `Endpoint` class, so it doesn't
use `convert_args` to convert organization slug. I added logic to handle
`organization_id_or_slug`. When I rename path parameters, I will modify
the function signature to `organization_id_or_slug`

wmak committed:
fix(trace): Handle multiple perf issues in a trace (#69473)

- This fixes a bug where we weren't handling multiple perf issues in a
single trace since we did it by mapping event_ids to occurrence_ids, but
there could be multiple occurrence_id for a given event_id
- This also fixes a bug (only with new spans view) where the list of
suspect spans was _all_ the suspect spans across all perf issues for
every perf issue, rather than just the suspect spans for a given perf
issue
- Snuck some decorators to add more tracing too

evanpurkhiser committed:
feat(crons): Remove crons-ownership feature flag (frontend) (#69521)

azaslavsky committed:
feat(relocation): Add API to list relocation's GCS artifacts (#69430)

This will be used in the admin panel exclusively by users with the
`relocation.admin` `UserPermission` set.

gggritso committed:
feat(perf): Remove "Starfish" sidebar section (#69524)

The time has come! The section is not needed, the stuff underneath it
needs to be moved out, etc.

gggritso committed:
fix(perf): Move "Give Feedback" button to header (#69527)

A lot of people are complaining that the floating button is blocking the
footer links. It's a more common pattern to put the button in the header
anyway, that's how the rest of the application does it.

dashed committed:
feat(slack): Debug function to generate slack block kit preview URLs (#69471)

Make it easier to generate Slack block kit preview URLs for
debugging/development:

<img width="1920" alt="Screenshot 2024-04-22 at 7 26 23 PM"
src="https://github.com/getsentry/sentry/assets/139499/9f749310-16f3-4adc-bfc2-8ef8501d0d5c">

This was inspired by:
https://github.com/raycharius/slack-block-builder/blob/fac1baf4054ebfa5241d6065db54858f1522fe7d/src/internal/methods/other-methods.ts#L68-L83

For example:
https://app.slack.com/block-kit-builder/T024ZCV9U#%7B%22blocks%22:%5B%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22*You've%20reached%2080%25%20of%20your%20performance%20unit%20quota%20limit*%20%20%5CnYour%20organization%20*baz*%20has%20consumed%2080%25%20of%20its%20%3Chttps://docs.sentry.io/product/performance/transaction-summary/%7Cperformance%20unit%3E%20capacity.%20We'll%20drop%20performance%20units%20when%20you%20exceed%20your%20quota.%22%7D%7D,%7B%22type%22:%22context%22,%22elements%22:%5B%7B%22type%22:%22mrkdwn%22,%22text%22:%22You%20are%20receiving%20this%20notification%20because%20you%20have%20the%20scope%20org:billing%20%7C%20%3Chttp://baz.testserver/settings/account/notifications/quota/?referrer=reserved_quota_threshold-slack-user&notification_uuid=fc0f0579-6133-400a-8c3b-7002c27749f7%7CNotification%20Settings%3E%22%7D%5D%7D,%7B%22type%22:%22actions%22,%22elements%22:%5B%7B%22type%22:%22button%22,%22text%22:%7B%22type%22:%22plain_text%22,%22text%22:%22Increase%20budget%22%7D,%22url%22:%22http://baz.testserver/settings/billing/checkout/?referrer=reserved_quota_threshold-slack-user&notification_uuid=fc0f0579-6133-400a-8c3b-7002c27749f7#step2%22,%22value%22:%22link_clicked%22%7D,%7B%22type%22:%22button%22,%22text%22:%7B%22type%22:%22plain_text%22,%22text%22:%22View%20Consumption%22%7D,%22url%22:%22http://baz.testserver/settings/billing/overview/?referrer=reserved_quota_threshold-slack-user&notification_uuid=fc0f0579-6133-400a-8c3b-7002c27749f7&category=transactions%22,%22value%22:%22link_clicked%22%7D%5D%7D%5D%7D

azaslavsky committed:
feat(relocation): Allow relocation artifacts to be downloaded (#69444)

If the relocation artifact is a `.tar` file, it is first decrypted.

This ability is only available to Superusers/staff who have the
`relocation.admin` permission.

JonasBa committed:
feat(trace): bring back dashed backgrounds (#69321)

scttcper committed:
test(ui): Prevent flake in sudo modal test (#69533)

lobsterkatie committed:
fix(ingest): Repair broken title data (#69398)

Last week, we ran into a bug wherein we were mistakenly classifying error events as default (message) events. Because the two types of events have slightly different schema, that meant that we were looking in the wrong place for title data, and as a result ended up with a whole mess of events all titled `<unlabled event>`. Because groups update their title with each new event that comes in, that meant we also had a bunch of groups called `<unlabeled event>`, with the result that the issue stream became an unusable sea of indistinguishable issues.

We quickly realized the problem and fixed the bug, but while that meant events started coming in with correct titles again, the groups weren't fully updating themselves. Why? Because it turns out that default events add title information to `group.data["metadata"]` as well as to `group.data` itself, where as error events only add it to `group.data`. This meant that as new events came in only half of the bad title data was getting overwritten by good title data, while the bad data in `group.data["metadata"]` was untouched. 

This fixes that by allowing `None` to overwrite existing title data, provided the existing title data is `<unlabeled event>` or one of its cousins (`<unknown>`, `<untitled>`, etc.) It also aims to prevent a similar thing from happening in the future, by not allowing `<unlabeled event>` (et al) to overwrite either a real title or a `None`. The same fix is applied to `group.message`, which comes in part from the title.

Note: To determine if a title is worthwhile, I reused the collection of titles we use when checking if an event's message is worth sending to seer or not. To make it a little clearer what they're about, I renamed the constant to `PLACEHOLDER_TITLES`, and to make things ever-so-slightly more efficient, I changed it from a list to a set.

anonrig committed:
feat: enable typescript biome rules (#69534)

We weren't leveraging enough Typescript rules. Adding these caught
some...

---------

Co-authored-by: getsantry[bot] <66042841+getsantry[bot]@users.noreply.github.com>

JoshFerge committed:
fix(replays): dont let superusers affect viewed by (#69529)

- dont show superusers user_ids in the viewed by response
- Don't create a viewed by record if the user is not in the org of the
replay

anonrig committed:
fix: disable prettier logging (#69536)

Disables useless prettier logs

scttcper committed:
feat(billing): Allow passing reserved, ondemand to usage chart (#69366)

markstory committed:
chore(actor) Backfill rule.owner fields (#69510)

As part of the work to remove `Actor` we need to populate
`owner_user_id` and `owner_team_id` on `Rule`.

New writes were addressed with #69427

oioki committed:
feat(integrations): Add confirmation modal to token revocation and clientSecret rotation (#69494)

These actions could be disruptive, so let's add one more confirmation
check.

* Token revocation on the Auth User Tokens page
* Token revocation and Client Secret rotation on the Edit
Internal/Public Integration page
* Client Secret rotation on the API Application Details page



https://github.com/getsentry/sentry/assets/1127549/6ed3d0ad-8760-4d87-8837-d262e13dcf42


https://github.com/getsentry/sentry/assets/1127549/62f6bb93-abec-4aab-bec3-b693267ec644


https://github.com/getsentry/sentry/assets/1127549/cac4c0e7-194b-4168-ac0c-0edb67195763

Also adding the last characters to the confirmation modal on the
Organization Auth Tokens page to make sure we revoke the correct one.


![image](https://github.com/getsentry/sentry/assets/1127549/f66771ec-f3e1-442e-b10e-99b15eb25ee6)

Zylphrex committed:
feat(trace-explorer): Support additional queries for separate spans (#69535)

This adds support for multiple queries on a trace allowing searches to
find traces with conditions on multiple spans.

iamrajjoshi committed:
feat(apigateway): Added proxy support for `sentry_app_id_or_slug` path params (#69530)

* Similar to https://github.com/getsentry/sentry/pull/69245

This PR adds support for `sentry_app_id_or_slug` existing in path
parameter, so when we go change all our apis from `sentry_app_slug` to
`sentry_app_id_or_slug`, we can still proxy correctly.

I also fixed a small type mismatch issue I saw with the value of
`org_id_or_slug`.

nhsiehgit committed:
disable monitorType on edit & update hover states & modify serializer… (#69541)

… to not allow update

![Screenshot 2024-04-23 at 2 18
28 PM](https://github.com/getsentry/sentry/assets/6186377/7e1d5f72-be0c-4889-91e9-5bf882e1691a)

evanpurkhiser committed:
ref(crons): Remove crons-ownership feature flag (backend) (#69522)

wedamija committed:
chore(post-process): Remove `Event.build_group_events` (#69472)

This isn't necessary anymore - events can have at most one group, and so
we can use `Event.for_group` instead. We'll probably continue
simplifying this more.

john-z-yang committed:
test(metrics): add more features to send_metrics.py (#69545)

### Overview

Adds a few more features to the send_metrics.py script.
- Allow b64 encoding of sets and distributions
- Allows user to specify which metrics type they would like to send

malwilley committed:
fix(autofix): Fix behavior and rendering of custom root cause response (#69551)

- Add handler to the "Continue with this fix" button in the custom
response section, which was missing it before
- Add padding and adjust wording to render the selected fix better
- Add tests

antonpirker committed:
Update Sentry Python SDK to version 2.0.0rc6. (#69496)

Updates Python SDK rc6. 
This new rc6 of the Python SDK fixed the problem of accumulating
`baggage` headers in the Celery integration (that was introduced in
2.0.0rc5).

obostjancic committed:
feat(stats): metrics stats group by project (#69488)

jjbayer committed:
fix(relay): Temporarily disable passive limits [INC-730] (#69564)

Passive projects do not enforce the cardinality limit, so it has no
product impact to disable them entirely.

ArthurKnaus committed:
ref(cardinality-limit): Streamline option names for cardinality limits (#69563)

The options for cardinality limits used the old `performance` and
`releasehealth` names instead of the regular namespace identifiers
`transactions` and `sessions`, respectively. Options with the correct
naming had been registered before.

This PR swaps out the mapping to use the correct naming and deprecates
the old options.

Requires https://github.com/getsentry/sentry-options-automator/pull/1210

loewenheim committed:
perf: Experimentally use orjson in Slack integration (#69409)

This adds an option `"integrations.slack.enable-orjson"` to enable
`orjson` for json parsing (ETA: and dumping) in the slack integration.

Ref: https://github.com/getsentry/sentry/issues/68903

armenzg committed:
feat(related_issues): Trace connected errors (#69237)

Given a group, we look for the recommended event and search for any
other errors in the trace.

These trace-connected groups will be shown under the Related Issues tab.
See the current look (more UI changes will be needed).

<img width="933" alt="image"
src="https://github.com/getsentry/sentry/assets/44410/b3f774d7-dc71-40b9-b144-65f143b6e981">

armenzg commented on src/sentry/search/events/builder/discover.py:
```diff
@@ -583,6 +584,10 @@ def resolve_params(self) -> list[WhereType]:
 
+        # This will prevent calling Snuba with an empty list of projects, thus, returning no data.
+        if not self.params.project_ids:
+            raise UnqualifiedQueryError("You need to specify at least one project.")
```
@wedamija @Zylphrex I'm now raising an error instead of relying on Snuba telling us.

I expect some backends to fail and I will adjust them.
